src/transform/src/map_lifting.rs

```diff
@@ -150,23 +203,54 @@ impl LiteralLifting {
-                        // Discard literals that are not projected.
-                        // TODO(frank): this could also discard intermediate
-                        // literals that are not projected, with more thought.
-                        while *project_max + 1 < input_arity + literals.len()
-                            && !literals.is_empty()
+                    // For each input literal contains a vector with the `output` positions
+                    // that references it. By putting data into a Vec and sorting, we
+                    // guarantee a reliable order.
+                    let mut used_literals = outputs
+                        .iter()
+                        .enumerate()
+                        .filter(|(_, x)| **x >= input_arity)
+                        .map(|(out_col, old_in_col)| (old_in_col - input_arity, out_col))
+                        // group them to avoid adding duplicated literals
+                        .into_group_map()
+                        .drain()
+                        .collect::<Vec<_>>();
+
+                    if used_literals.len() != literals.len() {
+                        used_literals.sort();
+                        // Discard literals that are not projected
+                        literals = used_literals
+                            .iter()
+                            .map(|(old_in_col, _)| literals[*old_in_col].clone())
+                            .collect::<Vec<_>>();
+                        // Update the references to the literal in `output`
+                        for (new_in_col, (_old_in_col, out_cols)) in
+                            used_literals.iter().enumerate()
                         {
-                            literals.pop();
+                            for out_col in out_cols {
+                                outputs[*out_col] = input_arity + new_in_col;
+                            }
                         }
                     }
+
                     // If the literals need to be re-interleaved,
                     // we don't have much choice but to install a
                     // Map operator to do that under the project.
                     // Ideally this doesn't happen much, as projects
                     // get lifted too.
                     if !literals.is_empty() {
-                        **input = input.take_dangerous().map(literals);
+                        let mut scalars = literals;
+                        // Merge nested Map operators.
+                        while let MirRelationExpr::Map {
+                            input: inner_input,
+                            scalars: inner_scalars,
+                        } = &mut **input
+                        {
+                            inner_scalars.extend(scalars.clone());
+                            scalars = inner_scalars.clone();
+                            **input = inner_input.take_dangerous();
+                        }
```

Comment:
Yes, that works. My reasoning for adding it in the first place was that when possible transformations should always be self-contained and should leave the plan in a more-optimal state. Assuming that less operators is always more optimal, that assumption is not true for the following input construct:

```
..
Map (#1 + 1), 1
Project (#2, #3)
```

which results in the following _less optimal_ output graph:

```
..
Map (#1 + 1)
Map 1
Project (#2, #3)
```

Reply:
I've removed the code as suggested. However, in the case above, the addition of an extra Map operator could be considered as an undesired side-effect of the transform, since it's meant to propagate the knowledge about literals to places where it may be useful, but in that case it's just splitting a `Map` operator.

I'd be inclined to put it in a shared method, or replace the WHILE with an IF so that it puts back the literals in the original `Map` leaving it as it was originally when reaching a `Project`. My reasoning being that assuming each transformation in a fixed point involves an entire traversal of the graph, the fewer of them the better. Specially when the need for adding a new transform to the list is just for cleaning up after some previous transform.

Reply:
> My reasoning being that assuming each transformation in a fixed point involves an entire traversal of the graph, the fewer of them the better. Specially when the need for adding a new transform to the list is just for cleaning up after some previous transform.

Indeed, less graph traversal would be better. I definitely think we should transition from the current paradigm, namely "Every transform should do distinct, non-overlapping stuff, and if your transform needs cleaning up, add a new transform to the list" to  "When possible transformations should always be self-contained and should leave the plan in a more-optimal state." 

However, there are two blockers in my opinion:
1) We as an organization have lost track of how our existing transforms are supposed to overlap/relate/interact with each other. Surveying this has been an ongoing project of mine for the past couple of months, and I prefer we not add more overlaps until I finish the survey. You can see part of the work in #6485, where I've been trying to rearrange the order of the transforms in more easily-describable and understood manner.  
2) We don't have a general framework yet for making each transform self-contained. Besides `LiteralLifting`, a bunch of other transforms could benefit from running cleanup actions during the same graph traversal. I would prefer for us to design a common cleanup template and then apply it every transform in need of cleanup instead of making custom cleanup actions every time we happen to touch a particular transform. I have created issue #6542 to track this. Feel free to add comments and ideas to the ticket!